pkg/csi/service/vanilla/controller.go

```diff
@@ -1349,10 +1351,208 @@ func (c *controller) ValidateVolumeCapabilities(ctx context.Context, req *csi.Va
 
 func (c *controller) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (
 	*csi.ListVolumesResponse, error) {
+	start := time.Now()
 	ctx = logger.NewContextWithLogger(ctx)
 	log := logger.GetLogger(ctx)
-	log.Infof("ListVolumes: called with args %+v", *req)
-	return nil, logger.LogNewErrorCode(log, codes.Unimplemented, "listVolumes")
+	volumeType := prometheus.PrometheusUnknownVolumeType
+
+	if !commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.ListVolumes) {
+		return nil, logger.LogNewErrorCode(log, codes.Unimplemented, "List Volumes")
+	}
+
+	cfg, err := common.GetConfig(ctx)
+	if err != nil {
+		return nil, logger.LogNewErrorf(log, "failed to read config. Error: %+v", err)
+	}
+
+	// Get the global query limit
+	maxEntries := cfg.Global.QueryLimit
+	if req.MaxEntries != 0 {
+		maxEntries = int(req.MaxEntries)
+	}
+
+	listVolumesInternal := func() (*csi.ListVolumesResponse, string, error) {
+		log.Infof("ListVolumes: called with args %+v", *req)
+
+		startingToken := 0
+		if req.StartingToken != "" {
+			startingToken, err = strconv.Atoi(req.StartingToken)
+			if err != nil {
+				log.Errorf("Unable to convert startingToken from string to int err=%v", err)
+				return nil, csifault.CSIInvalidArgumentFault, logger.LogNewErrorCode(log, codes.InvalidArgument,
+					"startingToken not a valid integer")
+			}
+		}
+
+		// Step 1: Get all the volume IDs of PVs, from K8s cluster
+		volIDsInK8s := commonco.ContainerOrchestratorUtility.GetAllK8sVolumes()
+		log.Debugf("Number of Volume IDs of PVs from K8s cluster %v, list of volumes %v", len(volIDsInK8s),
+			volIDsInK8s)
+
+		// Step 2: Get all Volume IDs from CNS QueryAll API
+		queryFilter := cnstypes.CnsQueryFilter{
+			ContainerClusterIds: []string{cfg.Global.ClusterID},
+		}
+		querySelection := cnstypes.CnsQuerySelection{
+			Names: []string{
+				string(cnstypes.QuerySelectionNameTypeVolumeType),
+			},
+		}
+		// Select only the volume type.
+		queryResult, err := c.manager.VolumeManager.QueryAllVolume(ctx, queryFilter, querySelection)
```

Comment:
why do we need call GetAllK8sVolumes and GetAllK8sVolumes again if we want to get the next set of entries in already populated list.